### DISCUSSION TITLE
iOS build?

### DISCUSSION THREAD

**Discussion by xloem** [markqvist/Reticulum#117] [Requests for New Features]
Hey, your software looks great. I’d love to try it on this iPad. It looks like Kivy can build to iOS; I haven’t tried this myself.

https://kivy.org/doc/stable/guide/packaging-ios.html

**Comment by markqvist** [comment]
Hi @xloem

I feel it is worth expanding a little on the facts and situation here, to explore why there is not, and probably will not be an iOS build of Sideband.

Due to some very fundamental differences in the foundation of the iOS/iPadOS operating systems, creating an "iOS Sideband" is not something I currently prioritise very highly. It *is* most probably _possible_ to build Sideband for iOS, but the app would have to include a lot of changes, that would severely limit the functionality of the system, or use a mountain of weird hacks to offer something near the same functionality on iOS as on Android and desktop.

In iOS, Apple does not allow any sort of extended background services, or persistent and full access to networking devices (and defininely not USB ports, even Bluetooth APIs are quite limited). This makes it practically impossible to create true peer-to-peer systems on iOS, since everything needs to access things like background notifications and tasks through Apples APIs, which in turn means doing everything through *their* infrastructure, and backends residing on the Internet (and only those that they approve of).

As you can see, such a system is **not** peer-to-peer and would be heavily dependent on both the continued cooperation of Apple, and on the availability and trustworthyness of the Internet at large.

One of the main points of Reticulum, LXMF and Sideband is to allow people to create and access communications systems that are as independent, resilient and self-sovereign an possible, and this is near impossible within a system such as Apples, where real hardware access practically non-existent.

Another incredibly sore point in this, is the fact that for anyone to obtain the software through the Apple ecosystem, they are required by Apple to be identified through several layers of invasice KYC. It is simply not possible to load software onto an iOS device without identifying yourself, and thus linking both the physical device and the action of installing the software to your personal identity. I sincerely uphold this as batshit insane, and I have a very, very hard time finding any sort of justification for myself, as a developer, to support such a system.

While I really hate to say it, most Apple devices sold these days are *not computers*, they are very sophisticated thin-client interfaces to a tightly controlled consumption system.

All of that being said, I might take a swing at it somehow, when I have time to do it, since of course it would be great to give iOS users access to the whole Reticulum ecosystem, but it will need to be through some method that is out of the grasp of Apples control.

One method could be loading a PWA onto the iOS device from a small external device, running the full Reticulum+LXMF stack, but this is obviously a bit of a larger project.

I will transfer this issue to the Reticulum discussion board, since I would also love to hear other perspectives exist in this regard, and I think it will have some more visibility on the forum. Thanks for bringing it up!

**Comment by xloem** [comment]
Thanks for your explanation.

- I’m using Yggdrasil on iOS to access hosts on the Yggdrasil p2p network, from other applications. It seems to run a background network service fine, although I do not know how. The source is at https://github.com/yggdrasil-network/yggdrasil-ios .

- I get my apple devices second hand. They do ask for an apple account to use the app store etc, but I didn’t know this required KYC. I do expect the identity of each user is tracked regardless, via automated profiling.

Your explanation makes sense.

**Reply by markqvist** [reply]
> They do ask for an apple account to use the app store etc, but I didn’t know this required KYC

Maybe it varies from area to area, but last time I tried creating an account with Apple (just to download their Xcode developer tools), they required an active phone number for SMS verification. While some may say that is not *strictly speakin* KYC, a phone number is practically always tied to full identification, hence my labeling it as KYC. Even as a developer, I refuse to accept a system where I need corporate blessings to *compile code to binaries*, and a whole other series of blessings and incantations to then distribute the software to potential users. It is completely unnecessary and wholly synthetic, and of course only serves the interests of the people in control of the system.

Absolutely brilliantly engineered control structure though, hats off for that. Flawless implementation. Could not have designed it any better myself, had I been so inclined. Conversely, though, my entire body of work is dedicated towards using cryptography for the exact opposite purpose of that: Freedom of thought, expression, communication and association.

There is, as far as I know, no way to "freely" load software onto an iOS device without having a developer account, and self-signing the binaries as a test-build, and then loading them via the developer tools. Only other option seems for developers to use the "TestFlight" system, that allows distributing "testing builds" to users, but Apple controls the number of installs, and the length of time they can remain active, so it is really hard to use it as a proper way to distribute software, both for users and for developers. I am pretty sure that the builds are signed in a way that contains an expiry date as well, after which they will be unable to run on the devices they were installed on.

> I’m using Yggdrasil on iOS to access hosts on the Yggdrasil p2p network, from other applications. It seems to run a background network service fine, although I do not know how.

It is not so much that it is impossible to run a network service in the background, but to run one that is *persistent*. I am pretty sure that the Yggdrasil app/service will also be put to sleep quite quickly once the device screen is off, like all other apps.

If your software is just operating within the (now almost universal) client-server paradigm, where your device just connects to a *server* on the Internet whenever it wants to transact anything (whether those servers exist within the Yggdrasil overlay fabric, or the normal Internet), and all "peers" in reality use *servers* to relay their information, that model is really no problem: When your phone wakes back up, the network service wakes up too, as soon as an app requests data from it, a connection is restored or reconnected, and the user is none the wiser.

If you want to actually transact information *truly* peer-to-peer, like Sideband allows, this model does not work. You need some sort of persistent network access, that is not put to sleep by the operating system. Otherwise your device effectively drops of the network once the OS sleeps the network daemon process. Someone tries to send you a message or access a service you provide? Too bad, your device is not actually connected anymore, and you cannot be reached.

Reticulum restores one of the original revolutionary concepts of the Internet, *true end-to-end connectivity*, and direct peer-to-peer communication, that is not mediated through server relays (as is the case in 99.9% of the cases now, even in systems calling themselves peer-to-peer).

While it is admittedly a while ago, and things may have changed now, I have *trawled* the iOS developer documentation for a solid way to deploy *actual* networking services on iOS, that can break out of the server-client paradigm, and it was my conclusion that the entire ecosystem is designed, amongst other things, to very specifically disallow exactly that.

Also, thanks for bringing up these issues and questions, I really appreciate it, since it gives me an oppertunity to expand on some of the, admittedly not very straight-forward, perspectives around why Reticulum and Sideband are not viable on iOS.

Maybe someday Reticulum will end up shipped as a kernel module in Darwin, but I would not hold my breath for it ;)

**Reply by ghost** [reply]
> There is, as far as I know, no way to "freely" load software onto an iOS device without having a developer account, and self-signing the binaries as a test-build, and then loading them via the developer tools.

Take a look at https://altstore.io/

**Reply by cobraPA** [reply]
Yes, I think in the EU there have been some court cases to require third party distribution.  In the US this is not the case, as far as I know, although there may still be some jailbreak activity that would allow third party installation.

**Comment by cobraPA** [comment]
Greetings @markqvist, I was going to open another discussion post, but maybe this is the place for it.  I am approaching the point of possibly testing a iOS client with a few other people.  As I see above, it doesn't sound like you will directly support TestFlight or App store testing at Apple?  So I wanted to get your thoughts on allowing someone else to do testing or distribution.  Would you allow that or have guidelines you would like followed in doing so?  I see you have applied copyright to the Sideband code, so would want you to approve any path to testing or distribution.

**Reply by markqvist** [reply]
If you've gotten it to this point of functionality already, you have my support in launching a TestFlight setup to try it out with some people. I honestly don't know exactly how to best manage the distribution if you do succeed in making a functional iOS port.

Mainly, I don't have the capacity to maintain an iOS build too, so it would be something you'd be taking on your own shoulders ;)

But let's figure out something if/when we get there. If you get a TestFlight thing running, I'd like to try it out too, I think I can borrow an old iPad from someone to test it.

**Reply by cobraPA** [reply]
Sure, sounds good.  Last I looked, TestFlight supported 100 users per app, so we can open it up here in discussion for people who want to test or try it out.  Do you have any naming preferences?  I see there is already a SideBand app in Apple land.  It is a music mixing app of some sort.  I would probably not want to use your actual SideBand name initially anyway, as it would reserve it for the first Apple developer taking it.  Let me know if you have a preference, maybe something like SideBandOS (Open Source?) or something more obscure.

I have basic interfaces working with a forced (default file with extra interfaces) defined for foreground operation now.  For ease of testing, I think it only needs to get interface configuration working in the UI at this point.  Background operation is still an investigation.

**Reply by faragher** [reply]
I have an old iPhone 6 I can bring back to life if you need to see how far back it works.

**Reply by cobraPA** [reply]
Nice, sure, good to try... For Apple devices, the limit isn't usually how old the device, but what version of iOS it runs.  Initially, I think the requirement will be iOS 10 due to a Thread interface availability.  There might be other things we find in testing with versions starting after 10 though...

**Reply by faragher** [reply]
It might actually be on an older OS. I recall trying to buy a new phone, backing it up, needing to upgrade iTunes to restore it, iTunes rewriting my USB drivers and making HIDs not work on my computer, using Windows recovery to roll it back. needing to update the old phone, then having like half the battery life and grindingly slow operation because of the new OS. That was when I switched to Android.

So it's either at 9, 10, or 11.

**Reply by faragher** [reply]
I have been trying to remember to circle back for *three hours* but I keep forgetting what I'm doing. iOS 11.2. I managed to get it running and even remember my password.

Old enough they cut me off from system services. :D

**Reply by cobraPA** [reply]
I posted a initial Beta but it requires iOS 12... Let me see if I can go back to 11 or 10 also.  Are you able to install TestFlight?  If you are cutoff maybe this won't help you yet...

**Reply by faragher** [reply]
I am locked out of the app store. Honestly, this is the only use case for this phone at this time, so I'll see if I can get it to update. I am not confident in its speed or usability with a new iOS, but it was just gathering dust for 1700 days according to my "you haven't backed up!" notification.

**Reply by cobraPA** [reply]
Ok, either way, I think I was able to make the build work for iOS 11 also, so if you are able to pull it, let me know how it looks.

**Reply by faragher** [reply]
Sorry, they won't let me download Test Flight unless I have iOS 14 and 12.5.7 is as high as the hardware can go.

**Reply by cobraPA** [reply]
Oooh, sorry.  Yeah, they do eventually move the lower limits on stuff as it gets older.
There is some discussion from developers that if you can access a newer device or Mac to get TestFlight 'purchased' (it's free), then the older device will prompt to 'install a compatible version' and pull an older TestFlight that is still compatible with iOS 12.  Seems to be some funkiness with the allowed download rules and acquiring the app for the first time.

**Comment by cobraPA** [comment]
Beta announcement in new [discussion](https://github.com/markqvist/Reticulum/discussions/307)
